Rover: Add param to reset WP origin while recovering from BendyRuler

rishabsingh3003 opened the pull request:
This PR fixes https://github.com/ArduPilot/ardupilot/issues/11565

I have added a parameter (OA_RESET_ORGIN) in OA library to allow the user if they want to reset the Waypoint origin or not while recovering from Bendy Ruler.
If OA_RESET_ORIGIN is set 0, the behaviour will be as it is in master right now. 
![before](https://user-images.githubusercontent.com/36970042/82551756-6facc680-9b7e-11ea-847c-af0e8be3722a.png)


But if OA_RESET_ORGIN is set to 1, which is the default in this PR, the following happens:
![after](https://user-images.githubusercontent.com/36970042/82551764-72a7b700-9b7e-11ea-90ba-739deba3c840.png)

@rmackay9 please let me know if this is the approach you were looking for. Thanks!

IamPete1 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -57,9 +58,16 @@ const AP_Param::GroupInfo AP_OAPathPlanner::var_info[] = {
     // @Group: DB_
     // @Path: AP_OADatabase.cpp
-    AP_SUBGROUPINFO(_oadatabase, "DB_", 4, AP_OAPathPlanner, AP_OADatabase),
+    AP_SUBGROUPINFO(_oadatabase, "DB_", 5, AP_OAPathPlanner, AP_OADatabase),
```
You cant change this number, it will reset the params to default, leave this as 4 and add the new one in as 5

rishabsingh3003 replied:
I have fixed it. Thanks for the catch!

IamPete1 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -57,9 +58,16 @@ const AP_Param::GroupInfo AP_OAPathPlanner::var_info[] = {
+    // @Description: Enabling this would reset the origin of the waypoint to the present location while recovering from Object Avoidance. Disabling it will make the vehicle return to its original waypoint trajectory.   
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO("RESET_ORIGIN", 4, AP_OAPathPlanner, _wp_reset, OA_WP_RESET_DEFAULT),
```
Maybe a AP_int16 bitmask param? Then you can find 15 other things to switch on and off too!

rishabsingh3003 replied:
I have left it for now cause it doesn't make too mcuh sense to me but I will ask others and if everyone agrees, i'll do it. Thanks a lot for the review :)

rmackay9 replied:
I'd recommend the param be called _OPTIONS.

rmackay9 commented:
Looking pretty good.  Let's also hide the parameter from Copter users

rishabsingh3003 commented:
I have addressed the following issues as discussed:
- Converted the param to a bit mask as suggested by @IamPete1. Thanks!
- Gotten rid of the global variable that I had included initially, changed it to local as suggested by @rmackay9 
- the param has been renamed to OPTIONS and has been made more generic. Hopefully we can add more things in the feature.

IamPete1 commented on libraries/AC_Avoidance/AP_OAPathPlanner.h:
```diff
@@ -54,6 +54,14 @@ class AP_OAPathPlanner {
+        OA_OPTION_WP_RESET = (1 << 0),
+    };
+
+    uint16_t get_options_on_recovery() const { return _options;}
```
might be worth changing to just get_options(), no big deal tho, you can always change it when you add some more

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -26,6 +26,9 @@ extern const AP_HAL::HAL &hal;
 // parameter defaults
 const float OA_LOOKAHEAD_DEFAULT = 15;
 const float OA_MARGIN_MAX_DEFAULT = 5;
+#if APM_BUILD_TYPE(APM_BUILD_Rover)
```
we don't really need this if here but it's OK.. just for next time ..

rishabsingh3003 replied:
@rmackay9 I actually had to force push this one as a specific change later because Travis CI was failing , stating "unused variable"

rmackay9 commented:
Merged, thanks!  Also thanks to @IamPete1 for the review.

rishabsingh3003 commented:
Thanks again @rmackay9 @IamPete1!
